api/src/main/java/org/apache/iceberg/ExpireSnapshots.java

```diff
@@ -118,4 +118,16 @@ public interface ExpireSnapshots extends PendingUpdate<List<Snapshot>> {
    * @return this for method chaining
    */
   ExpireSnapshots cleanExpiredFiles(boolean clean);
+
+  /**
+   * Allows removal of unreachable partition specs as part of the expiration operation
+   *
+   * @param removeUnusedSpecs setting this to true will remove partition specs that are no longer
+   *     reachable by any snapshot
+   * @return this for method chaining
+   */
+  default ExpireSnapshots removeUnusedSpecs(boolean removeUnusedSpecs) {
```

Comment:
Should we make this API a more generic `removeUnusedTableMetadata`? This goes back to the previous discussion that removing schemas and partition specs, requires the same level of work that the implementation has to do so imo there's not much value in separating them and forcing a user to chain multiple methods. Generally if they want to remove unused specs, they probably also want to remove unused schemas as well.

Reply:
Also, while I think we generally try and avoid boolean arguments in APIs, this may be one case where it makes sense. Down the line, if we want to make this behavior the default and have a path for users to disable cleanup of specs/schemas, they can.

Reply:
cc @danielcweeks @RussellSpitzer @rdblue @nastra thoughts?

Reply:
I'm not sure we want to even allow the option to not do this. Is there a benefit to leaving a spec or schema in place if it is no longer in use?

That said, I would be fine with just having a "cleanMetadata(boolean cleanMetadata: True)"

Reply:
@RussellSpitzer I was thinking about that, it's true that there's no real benefit to keeping the spec/schema in place but the issue I'm thinking about is more REST + compatibility related. 

A bit ago, we added the ability to send remove spec updates to the server and if we change the snapshot expiration implementation to just always remove metadata in the implementation, servers may not be able to handle that yet since it was recently added as a possible update type and services would perhaps unnecessarily fail the commit as part of expiration since the service would say unsupported. 

In the current model a client would opt-in knowing that the service would support the spec removal.
There may be a different way to handle this though so we can keep it all implicit in the procedure though.

Reply:
Thanks @advancedxy ! I'm in favor of the client side API option for now, just had a comment on the name of it. 

I think there's an important discussion to be had on how REST servers can indicate to clients the specific supported update types, this can either be done through config endpoint or through the existing endpoints discovery mechanism where each update type is attached as a query fragment to the endpoint and clients just use that.
 
My opinion is that it's not worth blocking this on a REST protocol discussion since we can always just come back and deprecate this client side option and have it run as part of the default procedure implementation.